src/show.cpp

```diff
@@ -21,11 +21,11 @@ int main(int argc, char **argv) {
     // first try own file format
     geo::ShapePtr shape = geo::serialization::fromFile(filename);
 
-    if (!shape) {
+    if (shape->empty()) {
```

Comment:
You're silently merging failure to read anything, and successfully reading an empty world onto the same result.
This invalidates the comment at line 25, and also the text at line 29.

If eg a zero-height box is also discarded (and thus non-empty world files could also result in an empty world), you also merged 'invalid data' (for as far as a zero-height box is invalid, which I somewhat doubt) into a read failure. The person that encounters one of these cases is going to spend a lot of time how and where it doesn't work.

It's a bit more work to differentiate between such cases, but when you want to load a new file 5 minutes before the competition starts and it returns "failed read".....   good luck to you then.

Reply:
I didn't think of that.

I my opinion the use of an empty pointer for failure case is not ok, because it much easier to create a segfault. If you create a different test for checking if your shape is actually valid, you reduce the chances of segfaults.

Do you agree? Or is the use of empty pointers acceptable in your opinion.

Reply:
For any function that produces a result, you first decide all possible results of the call. The normal behaviour is usual trivial, the tricky parts are in deciding which non-normal behaviour is relevant to the caller. Secondly, you create a way to represent each case in the returned data. Also here, non-normal behaviour is more difficult to encode usually.  With that set up, you can _code the function_. Note that how you encode each possible case in data is fully free.

To _use the function_ successfully, the programmer coding the call to the function (weeks to years later) has to know all possible results of the call, and how they are represented in returned data. In other words, there is a communication need between the author of a function, and the programmer using it in whatever code he/she makes. In general, this will happen several times for each function.

If you have the above communication under control, I think you can use any value to represent any result. The communication makes sure the user of the function is aware of all possible cases, and understands how to test the returned data to decide what the function is giving back.

As such, a crash due to not handling all possible return values from a function is thus not a software problem, it's the result of a failed communication.

As you say, you can avoid a crash by encoding a failure result in a different way in the returned data from the function. Instead of `null` you return an empty object, for example. Sure this avoids a crash at that point in the code, but it doesn't solve the failed communication. In other words, given that a `null` is not going to be properly communicated, how big is the chance that the caller of the function is going to understand to test for `x->empty()` before using the result?

A second consideration, suppose by shear luck, the caller does find your `x->empty()`, and decides to use it. Loaded file is empty. Now what? What's the next step to do? Have Amigo run in an empty world? Crash? Abort? Report an error and continue?

Is not crashing better than crashing?
You currently decided not crashing is less bad. That's a valid answer.

Ok, assuming Amigo in an empty world is not doing much good, you're going to have to debug this problem. So, what's better? A crash at the failed communication point, or an error message "help, I seem to be floating in free space, I am lost!!" quite likely in a totally different part of the code (eg the planner or ED that can't find something).

Both crashing and not crashing has benefits as well as costs in some way.